### DISCUSSION TITLE
sort + findFirstRow docs

### DISCUSSION THREAD

**Discussion by macMikey** [serverpod/serverpod#2814] [Q&A] [answered]
on the [crud page](https://docs.serverpod.dev/concepts/database/crud), there is this note:
<img width="834" alt="Screenshot 2024-10-03 at 22 51 04" src="https://github.com/user-attachments/assets/9f74ced6-149d-470f-acf7-97ea07e1498d">
if ordering is important, i guess that is supposed to mean that the sort will be applied before **findFirstRow**, but based on the wording, i am kind-of wondering if it is saying that the sort is superfluous, because only one record is selected.

**Comment by Isakdl** [comment]
If your query returns more than 1 record, the first record will be used. Postgres determine the first record and is not necessarily the insert order or deterministic. 

If you want to guarantee some ordering and that the ordering is always the same add an `orderBy` clause.

**Comment by macMikey** [comment]
i still don't understand.
the note is either:
* unnecessary and confusing, if the method only returns the first row of an unsorted query, sorting it doesn't make any difference, **or** 
* the note should be rewritten to make it clear that the selection will be sorted before it is cut down to a single row.

**Reply by Isakdl** [reply]
The only difference between this method and find, is that this method will take the first entry from the result. 

Calling `await Model.db.findFirstRow(...)` is equivalent to doing `(await Model.db.find(...)).firstOrNull;`

**Comment by macMikey** [comment]
thanks for being patient with me.
so, why the info box about sort, if it does not change any behavior?

**Reply by Isakdl** [reply]
It does change the behaviour as the order of the returned value may not be consistent over time.

Imagine you have a tabel with a column "name" that is a string. You have two entries in the database with the name "Mikey".

Now you do the following query:

```dart
final result = await User.db.findFirstRow(session, where: (t) => t.name.equals("Mikey"));
```

This query can return a different result each time you call it.

If you add the following
```dart
final result = await User.db.findFirstRow(session, where: (t) => t.name.equals("Mikey"), orderBy: (t) => t.id);
```

Then you will always get the same result, e.g. the one who happens to have the lowest id value.

This also holds true for a `find` the result order may be different on each query. 

In practice I believe postgres orders it by the update time but they make no guarantees around this.

**Comment by macMikey** [comment, marked as answer]
great. that's what i was asking.
one more time, to make sure: in **findFirstRow** with **orderBy**, the sequence is **find**, **orderBy**, **limit 1**, just like it was SQL, right?
if that's the case, then i'll clean up the info block to make that clear and issue a PR

**Reply by Isakdl** [reply]
Yes exactly, the entire query is in the end a normal SQL query.

**Comment by macMikey** [comment]
[PR 177](https://github.com/serverpod/serverpod_docs/pull/177)
